src/Processors/QueryPlan/Optimizations/mergeFilterIntoJoinCondition.cpp

```diff
@@ -0,0 +1,361 @@
+        /// After JOIN step these equalities will always evaluate to true.
+        for (const auto * & output : filter_dag.getOutputs())
+        {
+            auto it = conjuncts_to_replace.find(output);
+            if (it != conjuncts_to_replace.end())
+            {
+                output = &filter_dag.addColumn(ColumnWithTypeAndName(
+                    output->result_type->createColumnConst(1, 1),
+                    output->result_type,
+                    output->result_name));
+            }
+        }
+
+        if (rejected_conjuncts.size() == 1)
+        {
+            filter_dag.addOrReplaceInOutputs(filter_dag.addAlias(*rejected_conjuncts[0], filter_name));
+        }
+        else if (rejected_conjuncts.size() > 1)
+        {
+            FunctionOverloadResolverPtr func_builder_and = std::make_unique<FunctionToOverloadResolverAdaptor>(std::make_shared<FunctionAnd>());
+            filter_dag.addOrReplaceInOutputs(filter_dag.addFunction(func_builder_and, std::move(rejected_conjuncts), filter_name));
+        }
+        filter_dag.removeUnusedActions(/*allow_remove_inputs=*/false);
+    }
+
+    return { std::move(result), trivial_filter };
+}
+
+}
+
+size_t tryMergeFilterIntoJoinCondition(QueryPlan::Node * parent_node, QueryPlan::Nodes &  /*nodes*/, const Optimization::ExtraSettings &)
+{
+    if (parent_node->children.size() != 1)
+        return 0;
+
+    QueryPlan::Node * child_node = parent_node->children.front();
+
+    auto & parent = parent_node->step;
+    auto & child = child_node->step;
+
+    auto * filter_step = typeid_cast<FilterStep *>(parent.get());
+    auto * join_step = typeid_cast<JoinStepLogical *>(child.get());
+
+    if (!filter_step || !join_step)
+        return 0;
+
+    const auto & join_expressions = join_step->getExpressionActions();
+    auto & join_info = join_step->getJoinInfo();
+
+    if (join_info.kind == JoinKind::Full || join_info.kind == JoinKind::Left || join_info.kind == JoinKind::Right)
+        return 0;
+
+    auto strictness = join_info.strictness;
+    if (strictness == JoinStrictness::Anti || strictness == JoinStrictness::Asof || strictness == JoinStrictness::Any)
+        return 0;
+
+    const auto & join_header = child->getOutputHeader();
+    const auto & left_stream_header = child->getInputHeaders().front();
+    const auto & right_stream_header = child->getInputHeaders().back();
+
+    auto get_available_columns = [&join_header](const Block & input_header)
+    {
+        Names available_input_columns_for_filter;
+        const auto & input_columns_names = input_header.getNames();
+        available_input_columns_for_filter.reserve(input_columns_names.size());
+
+        for (const auto & name : input_columns_names)
+        {
+            if (!join_header.has(name))
+                continue;
+
+            /// Skip if type is changed. Push down expression expect equal types.
+            if (!input_header.getByName(name).type->equals(*join_header.getByName(name).type))
```

Comment:
Here is an unneeded `input_header.getByName(name)` lookup (if you iterate by input_header)